streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -57,16 +62,23 @@
 
     // must be maintained in topological order
     private final FixedOrderMap<String, Optional<StateStore>> registeredStores = new FixedOrderMap<>();
+    private final FixedOrderMap<String, Optional<StateStore>> globalStores = new FixedOrderMap<>();
 
     private final List<TopicPartition> changelogPartitions = new ArrayList<>();
 
     // TODO: this map does not work with customized grouper where multiple partitions
-    // of the same topic can be assigned to the same topic.
+    // of the same topic can be assigned to the same task.
     private final Map<String, TopicPartition> partitionForTopic;
 
+    private final boolean eosEnabled;
+    private final File baseDir;
+    private OffsetCheckpoint checkpointFile;
+    private final Map<TopicPartition, Long> checkpointFileCache = new HashMap<>();
+    private final Map<TopicPartition, Long> initialLoadedCheckpoints;
```

Comment:
Adding this collection breaks a circular dependency in this class:
* the checkpoints we load from disk are potentially not valid for the current topology
* we have to load the checkpoints immediately because we have to delete the checkpoint file before processing in the case of EOS
* we also need to have read the checkpoint file _before_ registering stores, since it might be needed to create a restorer
* we can't know if a checkpoint from the file is valid until _after_ registering stores 

In other words, if the prior code wanted to validate the loaded checkpoints, it would have to register the stores before loading checkpoints, but it also needs to load the checkpoints before registering the stores.

We're breaking the cycle here by keeping the loaded checkpoints separate. Now we read the checkpoint file into `initialLoadedCheckpoints`, which is used to register the stores, and then we are able to make sure that we only ever write valid checkpoints into the `checkpointFileCache`, which is used to update the checkpoint file later on.